In [1]:
pip install pulp

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.4/16.4 MB 98.2 MB/s eta 0:00:00


In [4]:
import yfinance as yf

In [5]:
import pandas as pd

In [6]:
tickers = ["BTC-USD", "CL=F"]

In [7]:
start_date = "2024-01-01"
end_date = "2024-12-31"

In [8]:
data = yf.download(tickers, start=start_date, end=end_date)
display(data.head())

/tmp/ipython-input-551500584.py:1: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download(tickers, start=start_date, end=end_date)
[*********************100%***********************]  2 of 2 completed


Price              Close                     High                      Low  \
Ticker           BTC-USD       CL=F       BTC-USD       CL=F       BTC-USD   
Date                                                                         
2024-01-01  44167.332031        NaN  44175.437500        NaN  42214.976562   
2024-01-02  44957.968750  70.379997  45899.707031  73.639999  44176.949219   
2024-01-03  42848.175781  72.699997  45503.242188  73.230003  40813.535156   
2024-01-04  44179.921875  72.190002  44770.023438  74.000000  42675.175781   
2024-01-05  44162.691406  73.809998  44353.285156  74.239998  42784.718750   

Price                          Open                  Volume            
Ticker           CL=F       BTC-USD       CL=F      BTC-USD      CL=F  
Date                                                                   
2024-01-01        NaN  42280.234375        NaN  18426978443       NaN  
2024-01-02  70.059998  44187.140625  71.709999  39335274536  330992.0  
2024-01-03  69.279999  44961.601562  70.500000  46342323118  334861.0  
2024-01-04  71.059998  42855.816406  73.040001  30448091210  344467.0  
2024-01-05  72.209999  44192.980469  72.400002  32336029347  325525.0

The data for "BTC-USD" and "CL=F" from January 1, 2024, to December 31, 2024, has already been downloaded and displayed in the previous cell using `yf.download()`.

In [10]:
close_prices = data['Close']
daily_returns = close_prices.pct_change()*100

daily_returns = daily_returns.dropna()
display(daily_returns.head())

/tmp/ipython-input-1715217252.py:2: FutureWarning: The default fill_method='pad' in DataFrame.pct_change is deprecated and will be removed in a future version. Either fill in any non-leading NA values prior to calling pct_change or specify 'fill_method=None' to not fill NA values.
  daily_returns = close_prices.pct_change()*100


Ticker       BTC-USD      CL=F
Date                          
2024-01-03 -4.692812  3.296391
2024-01-04  3.108058 -0.701506
2024-01-05 -0.039001  2.244071
2024-01-06 -0.392857  0.000000
2024-01-07 -0.104793  0.000000

In [11]:
import statsmodels.api as sm

In [13]:
Y = daily_returns["BTC-USD"]
X = daily_returns["CL=F"]

X = sm.add_constant(X)

model = sm.OLS(Y, X).fit()

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                BTC-USD   R-squared:                       0.006
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     2.057
Date:                Wed, 24 Sep 2025   Prob (F-statistic):              0.152
Time:                        04:18:18   Log-Likelihood:                -887.85
No. Observations:                 363   AIC:                             1780.
Df Residuals:                     361   BIC:                             1787.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2401      0.147      1.634      0.1

#predictive model

# Task
Analyze the relationship between BTC-USD and CL=F using a Random Forest regression model. Download historical data for both assets from Yahoo Finance for the period 2024-01-01 to 2024-12-31. Use CL=F as the predictor (X) and BTC-USD as the target (Y). Split the data into 80% training and 20% testing sets. Train a Random Forest model on the training data and evaluate its performance on the test data using RMSE and MAE metrics.

## Prepare data

### Subtask:
Define feature (X) and target (Y) variables for the model.


**Reasoning**:
Extract the target and feature variables from the `daily_returns` DataFrame according to the instructions.



In [14]:
Y = daily_returns["BTC-USD"]
X = daily_returns["CL=F"]

## Split data

### Subtask:
Split the data into training and testing sets (80% train, 20% test).


**Reasoning**:
Split the data into training and testing sets using an 80/20 ratio.



In [15]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

print("Training set shape:")
print("X_train:", X_train.shape)
print("Y_train:", Y_train.shape)
print("\nTesting set shape:")
print("X_test:", X_test.shape)
print("Y_test:", Y_test.shape)

Training set shape:
X_train: (290,)
Y_train: (290,)

Testing set shape:
X_test: (73,)
Y_test: (73,)


## Train model

### Subtask:
Train a Random Forest Regressor model on the training data.


**Reasoning**:
Import the Random Forest Regressor and train the model.



In [16]:
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(random_state=42)
model.fit(X_train.values.reshape(-1, 1), Y_train)

RandomForestRegressor(random_state=42)

## Make predictions

### Subtask:
Use the trained model to make predictions on the test data.


**Reasoning**:
Use the trained Random Forest model to predict the target variable on the test set.



In [17]:
Y_pred = model.predict(X_test.values.reshape(-1, 1))

## Evaluate model

### Subtask:
Calculate and display the Root Mean Squared Error (RMSE) and Mean Absolute Error (MAE) of the model's predictions.


**Reasoning**:
Calculate and display the RMSE and MAE of the model's predictions.



In [18]:
from sklearn.metrics import mean_squared_error, mean_absolute_error
import numpy as np

rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))
mae = mean_absolute_error(Y_test, Y_pred)

print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")

Root Mean Squared Error (RMSE): 4.04
Mean Absolute Error (MAE): 3.05


## Summary:

### Data Analysis Key Findings

*   The 'BTC-USD' daily returns were used as the target variable (Y), and 'CL=F' daily returns were used as the predictor variable (X).
*   The data was split into 80% for training (290 samples) and 20% for testing (73 samples).
*   A Random Forest Regressor model was trained on the training data.
*   The model's performance on the test set resulted in a Root Mean Squared Error (RMSE) of 4.04 and a Mean Absolute Error (MAE) of 3.05.

### Insights or Next Steps

*   The error metrics (RMSE and MAE) provide an indication of the model's prediction accuracy. Further investigation is needed to determine if these error values are acceptable for the specific application or if model tuning or alternative models are required.
*   Explore feature engineering or include additional relevant variables (e.g., other market indices, economic indicators) to potentially improve the model's predictive performance.
